### PR TITLE
Js support 2

### PR DESCRIPTION
**Pull request by tbuschto**

### CODE REVIEW

File: doc/databinding/index.md [hunk 1]
```diff
@@ -31,3 +33,19 @@ The following data binding related decorators are exported by `tabris-decorators
   * [@bindAll](./@bindAll.md) - Shorthand for `@bind({all})`
   * [@event](./@event.md) - Auto-initializes a [`Listeners`](../api/Listeners.md) properties to create a type-safe (change) event APIs.
   * [@getById](./@getById.md) - Auto-initializes a (usually private) custom component property for direct access to a child element.
+
+## Type Safety
+
+The data binding API can not provide compile-time type safety since it is based on ids and unchecked JSX attributes. However the framework will check at runtime whether the syntax is valid and points to existing elements and properties. If this is not the case an exception will be thrown when the bindings are established.
+
+The framework also attempts to check if the properties in a binding are type-safe. If they are not a "Unsafe binding" warning will be logged. This behavior can also be changed so that unsafe binding cause exceptions instead. To do so, simply add this code *before instantiating any widgets*:
```
Comment (cpetrov) [inline doc/databinding/index.md]: If they are not>,< a

File: doc/databinding/index.md [hunk 2]
```diff
@@ -31,3 +33,19 @@ The following data binding related decorators are exported by `tabris-decorators
...
+
+```js
+import {injector} from 'tabris-decorators';
+
+injector.jsxProcessor.unsafeBindings = 'error';
+```
+
+In TypeScript (`.ts`, `.tsx`) files unsafe binding only occur in certain edge-cases, assuming the `emitDecoratorMetadata` compiler option is enabled. In JavaScript they will be more common, especially in two-way bindings.
```
Comment (cpetrov) [inline doc/databinding/index.md]: files>,< unsafe binding only occur>s<

File: doc/databinding/index.md [hunk 3]
```diff
@@ -31,3 +33,19 @@ The following data binding related decorators are exported by `tabris-decorators
...
+
+To make an unsafe binding type-safe additional type information need to be provided to the [@property](./@property.md) and [@bind](./@bind.md) decorators. This may either be a [type guard](./@property.md#configtypeguard), or - in JavaScript - a the [`type`](./@property.md#configtype) option.
```
Comment (cpetrov) [inline doc/databinding/index.md]: type-safe>,< additional

File: doc/databinding/@bindAll.md [hunk 4]
```diff
@@ -4,6 +4,103 @@
...
+
+### Edge Cases
+
+The component property (`model` in the above example) may also be set to `null` (or `undefined`) at any time. In that case all the target properties of the child elements will be set back to their initial values. The initial value in this case refers to the value a target property had the moment the target element was attached to the component. The source properties on the former source object will retain their latest value.
+
+When a new two-way binding is established (when `_append` is called or the component property is assigned a new source object) all the target properties will be set to the current value of the their respective source property.
+
+There is one exception to this behavior: If the source property is set to `undefined` (but not `null`) at that moment it will be assigned the current value of the target property. If a source property is set to `undefined` later both properties are set to the initial value of the target property.
+
+If a source property converts or ignores the incoming value of the target property, the target property will follow and change again to contain the new source property value.
```
Comment (cpetrov) [inline doc/databinding/@bindAll.md]: You mean that the value will be converted by the source property's converter and the result will be passed to the target property's converter that will pass it to the native setter (that might change the value again)?

> the target property will follow and change again to contain the new source property value

I found the "will follow and change again ..." part confusing.

File: doc/databinding/@getById.md [hunk 5]
```diff
@@ -19,4 +81,8 @@ Makes the decorated property return the descendant with the same id as the prope
 
 Where `typeGuard` is of the type `value: any => boolean`.
 
-Like `@getById`, but uses the given type guard function to check the found widget, allowing widgets with a compatible API to be resolved as another type.
+Like `@getById`, but uses the given type guard function to check the found widget.
+
+It TypeScript this allows widgets with a compatible API to be resolved as another type.
```
[the remark this answers is not in the document]
Comment (cpetrov) [reply doc/databinding/@getById.md]: It TypeScript -> In TypeScript

File: doc/databinding/@getById.md [hunk 6]
```diff
@@ -6,7 +6,69 @@
 
 ## @getById (no parameter)
 
-Makes the decorated property return the descendant with the same id as the property name. The following rules apply:
+Makes the decorated property return a component-internal child element whose id is identical to the property name. This is useful if direct access to an internal widget is needed, e.g. to trigger an animation.
+
+> For simplicities sake the examples below append only a single child element, in which case there would be various simpler ways to obtain the reference. `@getById` is more useful when appended a larger widget tree.
```
[the remark this answers is not in the document]
Comment (cpetrov) [reply doc/databinding/@getById.md]: or "a larger widget tree is appended"

File: doc/databinding/index.md [hunk 7]
```diff
@@ -2,19 +2,21 @@
...
+Tabris provides [JSX](../JSX.md)/[decorators](http://www.typescriptlang.org/docs/handbook/decorators.html)-based declarative data binding via the `'tabris-decorators'` extension, which has to be installed in addition to the `tabris` module:
 
 ```
 npm install tabris-decorators
 ```
 
- [**This extension works only with TypeScript/JSX projects**](../typescript.md#setup). Mixed projects work as well, but the modules using the extension need to be `.ts` or `.tsx` files. It's strongly recommended to familiarize yourself with the [basic decorators syntax](https://www.typescriptlang.org/docs/handbook/decorators.html#class-decorators) if you aren't already.
+ [This extension is optimized for TypeScript/JSX projects](../typescript.md#setup), but JavaScript projects work as well as long is they use the TypeScript compiler to provide JSX and decorators support. It's strongly recommended to familiarize yourself with the [basic decorators syntax](https://www.typescriptlang.org/docs/handbook/decorators.html#class-decorators) if you haven't already.
```
Comment (cpetrov) [inline doc/databinding/index.md]: Maybe provide information what needs to be configured in the tsconfig.json or a minimal tsconfig.json?

File: src/decorators/bind.ts [hunk 8]
```diff
@@ -40,35 +40,44 @@ export type BindAllDecorator<ValidKeys extends string> = <
...
- * ‍@bind(path: '#childId.property')
- * componentProperty: SomeType = initialValue;
+ * ```ts
+ * ‍@bind('#childId.text')
+ * myProp: string = 'foo';
  * ```
  *
- * It's also possible to bind to properties of the object in the component property:
+ * *Notes:*
+ * * *`@bind` behaves like `@property` in most regards.*
```
Comment (cpetrov) [inline src/decorators/bind.ts]: Is this being displayed in an expected way in VSCode? Maybe use another character for bullets, e.g. `-` or `+` to make it easier to read and don't emphasize the bullet point text for simplicity.

Comment also applies to the other occurrences of this pattern.

Comment (tbuschto) [reply src/decorators/bind.ts]: Eh, it just needs to look good in VSC.

File: doc/databinding/index.md (hunk 7 above)
Comment (tbuschto) [reply doc/databinding/index.md]: The link provides this.

**Pull request merged**